qcore/gsf.py

```diff
@@ -0,0 +1,211 @@
+    y_bottom : np.ndarray
+        Coordinates of the bottom y boundary (lat, lon, depth).
+    resolution : float
+        Resolution of the meshgrid.
+
+    Returns
+    -------
+    np.ndarray
+        The meshgrid of the rectangular planar region.
+    """
+    origin = coordinates.wgs_depth_to_nztm(origin)
+    x_upper = coordinates.wgs_depth_to_nztm(x_upper)
+    y_bottom = coordinates.wgs_depth_to_nztm(y_bottom)
+    length_x = np.linalg.norm(x_upper - origin)
+    length_y = np.linalg.norm(y_bottom - origin)
+    nx = gridpoint_count_in_length(length_x, resolution)
+    ny = gridpoint_count_in_length(length_y, resolution)
+    x = np.linspace(0, length_x, nx)
+    y = np.linspace(0, length_y, ny)
+    xv, yv = np.meshgrid(x, y)
+    subdivision_coordinates = np.vstack([xv.ravel(), yv.ravel()])
+    transformation_matrix = np.vstack(
+        [(x_upper - origin) / length_x, (y_bottom - origin) / length_y]
+    ).T
+    nztm_meshgrid = (transformation_matrix @ subdivision_coordinates).T
+    nztm_meshgrid += origin
+    return coordinates.nztm_to_wgs_depth(nztm_meshgrid)
+
+
+def write_fault_to_gsf_file(
+    gsf_filepath: Path,
+    meshgrids: list[np.ndarray],
+    lengths: np.ndarray,
+    widths: np.ndarray,
+    strikes: np.ndarray,
+    dips: np.ndarray,
+    rakes: np.ndarray,
+    resolution: int,
+):
+    """Writes geometry data to a GSF file.
+
+    Parameters
+    ----------
+    gsf_filepath : Path
+        The file path pointing to the GSF file to write to.
+    meshgrids : list[np.ndarray]
+        List of meshgrid arrays representing fault segments (length: n).
+    lengths : np.ndarray
+        An (n x 1) array of segment lengths.
+    widths : np.ndarray
+        An (n x 1) array of segment widths.
+    strikes : np.ndarray
+        An (n x 1) array of segment strikes.
+    dips : np.ndarray
+        An (n x 1) array of segment dips.
+    rakes : np.ndarray
+        An (n x 1) array of segment rakes.
+    resolution : int
+        Resolution of the meshgrid.
+    """
+    with open(gsf_filepath, "w", encoding="utf-8") as gsf_file_handle:
+        gsf_file_handle.write(
+            "# LON  LAT  DEP(km)  SUB_DX  SUB_DY  LOC_STK  LOC_DIP  LOC_RAKE  SLIP(cm)  INIT_TIME  SEG_NO\n"
+        )
+        number_of_points = sum(meshgrid.shape[0] for meshgrid in meshgrids)
+        gsf_file_handle.write(f"{number_of_points}\n")
+        for i, (length, width, strike, dip, rake, meshgrid) in enumerate(
+            zip(lengths, widths, strikes, dips, rakes, meshgrids)
+        ):
+            strike_step = length / gridpoint_count_in_length(length * 1000, resolution)
+            dip_step = width / gridpoint_count_in_length(width * 1000, resolution)
+            for point in meshgrid:
+                gsf_file_handle.write(
+                    f"{point[1]:11.5f} {point[0]:11.5f} {point[2] / 1000:11.5e} {strike_step:11.5e} {dip_step:11.5e} {strike:6.1f} {dip:6.1f} {rake:6.1f} {-1.0:8.2f} {-1.0:8.2f} {i:3d}\n"
```

Comment:
This line is possibly a bit long? Black doesn't break up string lines, so has to be done manually.